src/Features/CSharp/Portable/Completion/KeywordRecommenders/StackAllocKeywordRecommender.cs

```diff
@@ -16,20 +16,48 @@ public StackAllocKeywordRecommender()
 
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
-            // type t = |
-            var token = context.TargetToken;
-            if (token.IsUnsafeContext())
+            var node = context.TargetToken.Parent;
+
+            // In case of an empty string
+            if (node == null)
+            {
+                return false;
+            }
+
+            // After a cast
+            if (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) &&
+                (node.IsKind(SyntaxKind.ParenthesizedExpression) || node.IsKind(SyntaxKind.CastExpression)))
+            {
+                node = node.Parent;
+            }
+
+            // Inside a conditional expression: value ? stackalloc : stackalloc
+            while (node.IsKind(SyntaxKind.ConditionalExpression) &&
+                (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) || context.TargetToken.IsKind(SyntaxKind.QuestionToken) || context.TargetToken.IsKind(SyntaxKind.ColonToken)))
+            {
+                node = node.Parent;
+            }
+
+            // assignment: x = stackalloc
+            if (node.IsKind(SyntaxKind.SimpleAssignmentExpression))
+            {
+                return node.Parent.IsKind(SyntaxKind.ExpressionStatement);
+            }
+
+            // declaration: var x = stackalloc
+            if (node.IsKind(SyntaxKind.EqualsValueClause))
             {
-                if (token.Kind() == SyntaxKind.EqualsToken &&
-                    token.Parent.IsKind(SyntaxKind.EqualsValueClause) &&
-                    token.Parent.IsParentKind(SyntaxKind.VariableDeclarator) &&
-                    token.Parent.Parent.IsParentKind(SyntaxKind.VariableDeclaration))
+                node = node.Parent;
+
+                if (node.IsKind(SyntaxKind.VariableDeclarator))
                 {
-                    var variableDeclaration = (VariableDeclarationSyntax)token.Parent.Parent.Parent;
-                    if (variableDeclaration.IsParentKind(SyntaxKind.LocalDeclarationStatement) ||
-                        variableDeclaration.IsParentKind(SyntaxKind.ForStatement))
+                    node = node.Parent;
+
+                    if (node.IsKind(SyntaxKind.VariableDeclaration))
                     {
-                        return variableDeclaration.Type.IsVar || variableDeclaration.Type.IsKind(SyntaxKind.PointerType);
+                        node = node.Parent;
+
+                        return node.IsKind(SyntaxKind.LocalDeclarationStatement) || node.IsKind(SyntaxKind.ForStatement);
```

Comment:
'var' is also allowed in a using.  I'm guessing we don't allow stackalloc there though because it would not be disposable?

Reply:
That's true.

---
In reply to: [156193066](https://github.com/dotnet/roslyn/pull/23655#discussion_r156193066) [](ancestors = 156193066)